File: 1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero.cpp
class Solution {
public:
    int minReorder(int n, vector<vector<int>>& connections) {
        vector<int> visited(n);
        vector<vector<int>> adj(n), back(n);
        int res = 0;
        queue<int> q;
        q.push(0);
        for (auto c : connections) {
            adj[c[0]].push_back(c[1]);
            back[c[1]].push_back(c[0]);
        }
        
        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            visited[curr] = 1;
            for (auto a: adj[curr]) {
                if (!visited[a]) {
                    res++;
                    q.push(a);
                }
            }
            for (auto b: back[curr]) {
                if (!visited[b]) {
                    q.push(b);
                }
            }
        }
        return res;
    }
};

File: 1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    private int count = 0;
    private Map<Integer, Boolean>[] adj;

    private void dfs(int city, int parent) {
        for (Integer nbr : adj[city].keySet()) {
            Boolean needReorder = adj[city].get(nbr);
            if (nbr != parent) {
                if (needReorder) {
                    count++;
                }
                dfs(nbr, city);
            }
        }
    }

    public int minReorder(int n, int[][] connections) {
        adj = new HashMap[n];
        Arrays.setAll(adj, map -> new HashMap<>());

        for (int[] conn : connections) {
            adj[conn[0]].put(conn[1], true);
            adj[conn[1]].put(conn[0], false);
        }
        dfs(0, -1);
        return count;
    }
}